File: frontend/src/services/api/novel-promotion/locations.ts
import type {
  NPAICreateLocationPayload,
  NPAIModifyLocationPayload,
  NPLocation,
  NPLocationCreate,
  NPLocationUpdate,
  NPSelectLocationImagePayload,
  NPTaskQueued,
} from '../../../types/novel-promotion'
import { apiClient } from '../client'

export async function listNPLocations(projectId: string) {
  const response = await apiClient.get(`/novel-promotion/${projectId}/locations`)
  return response.data.data.locations as NPLocation[]
}

export async function createNPLocation(projectId: string, payload: NPLocationCreate) {
  const response = await apiClient.post(`/novel-promotion/${projectId}/locations`, payload)
  return response.data.data.location as NPLocation
}

export async function updateNPLocation(
  projectId: string,
  locationId: string,
  payload: NPLocationUpdate,
) {
  const response = await apiClient.patch(
    `/novel-promotion/${projectId}/locations/${locationId}`,
    payload,
  )
  return response.data.data.location as NPLocation
}

export async function deleteNPLocation(projectId: string, locationId: string) {
  const response = await apiClient.delete(
    `/novel-promotion/${projectId}/locations/${locationId}`,
  )
  return response.data.data as { deleted: true }
}

export async function selectNPLocationImage(
  projectId: string,
  locationId: string,
  payload: NPSelectLocationImagePayload,
) {
  const response = await apiClient.post(
    `/novel-promotion/${projectId}/locations/${locationId}/select-image`,
    payload,
  )
  return response.data.data.location as NPLocation
}

export async function aiCreateNPLocation(projectId: string, payload: NPAICreateLocationPayload) {
  const response = await apiClient.post(
    `/novel-promotion/${projectId}/locations/ai-create`,
    payload,
  )
  return response.data.data as NPTaskQueued
}

export async function aiModifyNPLocation(
  projectId: string,
  locationId: string,
  payload: NPAIModifyLocationPayload,
) {
  const response = await apiClient.post(
    `/novel-promotion/${projectId}/locations/${locationId}/ai-modify`,
    payload,
  )
  return response.data.data as NPTaskQueued
}

File: backend/app/api/routes/np/locations.py
from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session

from app.api.deps.auth import get_current_user
from app.core.db import get_db
from app.db.models.user import User
from app.schemas.novel_promotion.entities import (
    AICreateLocationPayload,
    AIModifyLocationPayload,
    LocationCreate,
    LocationRead,
    LocationUpdate,
    SelectLocationImagePayload,
)
from app.services.novel_promotion.common import ensure_np_project
from app.services.novel_promotion.locations import LocationService
from app.services.novel_promotion.task_queue import queue_np_task

router = APIRouter()


def _ok(loc) -> dict:
    return {"success": True, "data": {"location": LocationRead.model_validate(loc).model_dump()}}


@router.get("/{project_id}/locations")
def list_locations(
    project_id: str,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
) -> dict:
    items = LocationService(db).list_(current_user.id, project_id)
    return {
        "success": True,
        "data": {"locations": [LocationRead.model_validate(it).model_dump() for it in items]},
    }


@router.post("/{project_id}/locations")
def create_location(
    project_id: str,
    payload: LocationCreate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
) -> dict:
    return _ok(LocationService(db).create(current_user.id, project_id, payload))


@router.patch("/{project_id}/locations/{location_id}")
def update_location(
    project_id: str,
    location_id: str,
    payload: LocationUpdate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
) -> dict:
    return _ok(LocationService(db).update(current_user.id, project_id, location_id, payload))


@router.delete("/{project_id}/locations/{location_id}")
def delete_location(
    project_id: str,
    location_id: str,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
) -> dict:
    LocationService(db).delete(current_user.id, project_id, location_id)
    return {"success": True, "data": {"deleted": True}}


@router.post("/{project_id}/locations/{location_id}/select-image")
def select_location_image(
    project_id: str,
    location_id: str,
    payload: SelectLocationImagePayload,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
) -> dict:
    return _ok(
        LocationService(db).select_image(
            current_user.id, project_id, location_id, payload.image_id
        )
    )


@router.post("/{project_id}/locations/ai-create")
def ai_create_location(
    project_id: str,
    payload: AICreateLocationPayload,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
) -> dict:
    np = ensure_np_project(db, current_user.id, project_id)
    task = queue_np_task(
        db,
        user_id=current_user.id,
        project_id=project_id,
        task_type="np_ai_create_location",
        target_type="np_project",
        target_id=np.id,
        payload=payload.model_dump(),
    )
    return {"success": True, "data": {"task_id": task.id}}


@router.post("/{project_id}/locations/{location_id}/ai-modify")
def ai_modify_location(
    project_id: str,
    location_id: str,
    payload: AIModifyLocationPayload,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
) -> dict:
    task = queue_np_task(
        db,
        user_id=current_user.id,
        project_id=project_id,
        task_type="np_ai_modify_location",
        target_type="np_location",
        target_id=location_id,
        payload=payload.model_dump(),
    )
    return {"success": True, "data": {"task_id": task.id}}
